core/src/main/scala/io/snappydata/SnappyTableStatsProviderService.scala

```diff
@@ -257,18 +257,20 @@ object SnappyEmbeddedTableStatsProviderService extends TableStatsProviderService
             // TODO: this should use a transactional iterator to get a consistent
             // snapshot (also pass the same transaction to getNumColumnsInTable
             //   for reading value and delete count)
-            val itr = new pr.PRLocalScanIterator(true /* primaryOnly */ , null /* no TX */ ,
+            val itr = new pr.PRLocalScanIterator(false /* primaryOnly */ , null /* no TX */ ,
               null /* not required since includeValues is false */ ,
               createRemoteIterator, false /* forUpdate */ , false /* includeValues */)
             // using direct region operations
             while (itr.hasNext) {
               val re = itr.next().asInstanceOf[AbstractRegionEntry]
               val key = re.getRawKey.asInstanceOf[ColumnFormatKey]
-              if (numColumnsInTable < 0) {
-                numColumnsInTable = key.getNumColumnsInTable(table)
+              if (itr.getHostedBucketRegion.getBucketAdvisor.isPrimary) {
+                if (numColumnsInTable < 0) {
+                  numColumnsInTable = key.getNumColumnsInTable(table)
+                }
+                rowsInColumnBatch += key.getColumnBatchRowCount(itr, re,
+                  numColumnsInTable)
               }
```

Comment:
Bit of perf optimization possible. ColumnFormatKey.getColumnBatchRowCount also does itr.getHostedBucketRegion. In fact that's the only reason it uses itr. Can you change getColumnBatchRowCount to be passed BucketRegion instead of PREntriesIterator and then get the BucketRegion only once from iterator? Will be cleaner API for ColumnBatchKey interface too.

Reply:
Sure. Will change it.